### PR TITLE
Add Missing Cases for Nullable Enums

### LINKED ISSUE #40240: 'Add missing cases' action when switching on nullable enum and bool
**Version Used**: 16.4

```cs
public static string Foo(Bar? bar)
{
    [||]switch (bar)
    {
    }
}
```

Expected action: 'Add missing cases'

```cs
public static string Foo(Bar? bar)
{
    switch (bar)
    {
        case Bar.Option1:
            break;
        case Bar.Option2:
            break;
        case null:
            break;
    }
}
```

### PR DESCRIPTION
**Pull request by 06needhamt**
Fixes #40240.

This PR fixes the 'add missing cases' action to add missing cases to switch statements where the switch expression type is a Nullable Enum.

In C# everything works as expected and the code action performs exactly the same with nullable and non-nullable enum types. However, in Visual Basic there is an issue because the Constant Expressions for the enum values are not being evaluated properly. I have confirmed that everything works as expected with non-nullable enum types. 

I have filed a separate issue to track the above at #40675. I'm not sure if it's related to the fix but I suspect its something else as it works as expected in C#

### CODE REVIEW

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ internal static class PopulateSwitchExpressionHelpers
         public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperation operation)
         {
             var switchExpression = operation.Value;
-            var switchExpressionType = switchExpression?.Type;
+            var switchExpressionType = (INamedTypeSymbol?)switchExpression?.Type;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: how do you know it's an INamedType here?  if verified earlier, can you add comment?

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 2]
```diff
@@ -28,6 +28,16 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
...
+            else if (switchExpressionType?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T &&
+                        switchExpressionType?.TypeArguments[0].TypeKind == TypeKind.Enum)
+            {
+                var underlyingEnum = switchExpressionType.TypeArguments[0];
+                if (!PopulateSwitchStatementHelpers.TryGetAllEnumMembers(underlyingEnum, enumMembers) ||
+                    !TryRemoveExistingEnumMembers(operation, enumMembers))
+                {
+                    return SpecializedCollections.EmptyCollection<ISymbol>();
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: same code as above.  instead, can you just do the nullable check first, and exract out the nullable type arg and then unilaterally do the enum chec?

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs [hunk 3]
```diff
@@ -54,6 +54,16 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchOperation switch
...
+            else if (switchExpressionType?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T &&
+                        switchExpressionType?.TypeArguments[0].TypeKind == TypeKind.Enum)
+            {
+                var underlyingEnum = switchExpressionType.TypeArguments[0];
+                if (!TryGetAllEnumMembers(underlyingEnum, enumMembers) ||
+                    !TryRemoveExistingEnumMembers(switchStatement, enumMembers))
+                {
+                    return SpecializedCollections.EmptyCollection<ISymbol>();
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs]: same issue.

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs (hunk 2 above)
Comment (06needhamt) [reply src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: I've updated the code to reflect your comments

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs (hunk 3 above)
Comment (06needhamt) [reply src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs]: As above

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs (hunk 1 above)
Comment (06needhamt) [reply src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: I've refactored the code so we only convert to an INamedTypeSymbol if the type is a named type.

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 4]
```diff
@@ -20,6 +20,14 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
             var switchExpressionType = switchExpression?.Type;
 
             var enumMembers = new Dictionary<long, ISymbol>();
+
+            if (switchExpressionType?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
+            {
+                // Check if the type of the expression is a nullable INamedTypeSymbol
+                // if the type is both nullable and an INamedTypeSymbol extact the type argument from the nullable
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]:
```suggestion
                // if the type is both nullable and an INamedTypeSymbol extract the type argument from the nullable
```

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 5]
```diff
@@ -20,6 +20,14 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
...
+                // and check if it is of enum type
+                switchExpression?.Type.IsNullable(out switchExpressionType);
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]:
```suggestion
            }

```

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs [hunk 6]
```diff
@@ -46,6 +47,15 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchOperation switch
             var switchExpressionType = switchExpression?.Type;
 
             var enumMembers = new Dictionary<long, ISymbol>();
+
+            if (switchExpressionType?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
+            {
+                // Check if the type of the expression is a nullable INamedTypeSymbol
+                // if the type is both nullable and an INamedTypeSymbol extact the type argument from the nullable
+                // and check if it is of enum type
+                switchExpression?.Type.IsNullable(out switchExpressionType);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchStatementHelpers.cs]:
```
switchExpressionType  = switchExpressionType.IsNullable(out underlyingType) ? underlyingType : switchExpressionType;
```

**Comment by 06needhamt** [comment]
@CyrusNajmabadi Fixed. Sorry about all the commits; I’m away from my computer and made the edits on mobile.

**Comment by CyrusNajmabadi** [comment]
Nothing to be sorry about! :)

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 7]
```diff
@@ -20,6 +20,15 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
             var switchExpressionType = switchExpression?.Type;
 
             var enumMembers = new Dictionary<long, ISymbol>();
+
+            if (switchExpressionType?.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: sorry, i wasn't clear.  you don't even need this 'if' part :)

Comment (06needhamt) [reply src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: No Problem I’ll remove it.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by 06needhamt** [comment]
Can I get a second review so we can move forward with this.

Thanks

**Comment by CyrusNajmabadi** [comment]
@ryzngard Are you back in the office this week?  If so, can you take a look at this community PR?  Thanks much!

**Comment by ryzngard** [comment]
@CyrusNajmabadi yup I'm back. Should be getting through a backlog of reviews shortly

**Comment by CyrusNajmabadi** [comment]
@ryzngard Awesome!  Thanks so much :)

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 8]
```diff
@@ -20,6 +20,12 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
             var switchExpressionType = switchExpression?.Type;
 
             var enumMembers = new Dictionary<long, ISymbol>();
+
+            // Check if the type of the expression is a nullable INamedTypeSymbol
+            // if the type is both nullable and an INamedTypeSymbol extract the type argument from the nullable
+            // and check if it is of enum type
+            switchExpressionType = switchExpressionType.IsNullable(out var underlyingType) ? underlyingType : switchExpressionType;
```
Comment (ryzngard) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: The line below was assuming `switchExpressionType` could be null, which is no longer checked here

Comment (06needhamt) [reply src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: Done

**Comment by ryzngard** [comment]
Integration test failure linked to #38198 , queuing a rerun

File: src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs [hunk 9]
```diff
@@ -20,6 +20,13 @@ public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchExpressionOperat
             var switchExpressionType = switchExpression?.Type;
 
             var enumMembers = new Dictionary<long, ISymbol>();
+
+            // Check if the type of the expression is a nullable INamedTypeSymbol
+            // if the type is both nullable and an INamedTypeSymbol extract the type argument from the nullable
+            // and check if it is of enum type
+            if (switchExpressionType != null)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/PopulateSwitch/PopulateSwitchExpressionHelpers.cs]: null check not needed.  `IsNullable` is null-safe (and we try to make all our `.Is` extensions that way).

**Pull request merged**